### ISSUE TITLE
[Java] Models with arrays of inner number types are wrongly, uncompiable generated

### ISSUE THREAD

**Issue by sdoeringNew** [swagger-api/swagger-codegen#6545]
##### Description

Models with arrays of inner number types are wrongly generated. Generally a number type with enum is generated as `NUMBER_0(0)`.
If this number type with enum is part of an array it is generated as: `_0("0")` The `NUMBER` is missing and the value would be generated as String.
This doesn't even compile.

##### Swagger-codegen version

2.2.3

##### Swagger declaration file content or url

```json
        "ModelWithArrayHavingNumbersWithEnum": {
            "type": "object",
            "properties": {
                "arrayWithNumbers": {
                    "type": "array",
                    "items": {
                        "type": "integer",
                        "format": "int32",
                        "enum": [0, 1, 2]
                    }
                }
            }
        }
```

##### Command line used for generation

..

##### Steps to reproduce

..

##### Related issues/PRs

#5712, #5619

##### Suggest a fix/enhancement

..

**Comment by wing328** [comment]
@sdoeringNew thanks for reporting the issue. You're right. I don't think we've the right code to handle array of enum yet.

Would you have time to contribute the enhancement? If yes, I will try to find some good starting points to fix this problem.

**Comment by sdoeringNew** [comment]
Hi Wing.

Yeah, sure. I can contribute a fix. In fact I already started ...until I
noticed there is currently no way to test this bug does not happen again.
How am I supposed to test that generated model files for Java will never
have that particular bug again?

Yes. There are the samples catching a lot. For example if they not compile
or a test fails there is probably something wrong with the mustaches.

I already wrote a test yaml with extensive enum usage. But there can I add
this file and generate the models in a testing scenario?
I don't think the samples are not the appropriate place for that as this is
no particular sample for a REST framework.

I thought of adding a new module for such kind of tests. To at least test
if a API generates and compiles successfully.

But I'm not in the position to decide those things. I'll do the tests for
the fix internally.

Kind regards
 Sven


Am 24.09.2017 17:22 schrieb "wing328" <notifications@github.com>:

**Comment by sdoeringNew** [comment]
After creating an test module with an extensive enum test API the following issues appeared:

- inner double / float / long / integer enums defined in an array are wrongly generated - they does not compile
- outer double enums wrongly generated - they does not compile
- outer double / float enums are differently generated as inner double / float enums - missing "_DOT_0"


I think some sort of tests are desperately needed. But where to put them?

**Comment by wing328** [comment]
@sdoeringNew please add them to https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml#L1382. Currently, we only have array of enum string but not array of enum of other types (integer, double, etc)

**Comment by sdoeringNew** [comment]
Okay.

But where do I put the tests to check that the generated Java models are as expected? Currently I only see Tests that the CodegenProperty are properly filled.

**Comment by wing328** [comment]
What about adding some tests in https://github.com/swagger-api/swagger-codegen/tree/master/samples/client/petstore/java/okhttp-gson/src/test/java/io/swagger/client? You will need to run "./bin/java-petstore-okhttp-gson.sh" or ".\bin\windows\java-petstore-okhttp-gson.bat" to update the Petstore samples.

**Comment by sdoeringNew** [comment]
Okay.

But what about the 69 other samples using this petstore-with-fake-endpoints-models-for-testing.yaml file? I don't believe that all of the non-Java projects will still work / compile / test. Do I need to fix those code generations, too? 🤕

**Comment by wing328** [comment]
Yes, we'll need to apply similar fix to other generators. Don't worry about those for the time being.

**Comment by sdoeringNew** [comment]
Consider it done.

Btw.: It was a good thing you proposed a GSON sample.
There were more bugs, too. 🤣 
- JsonReader#nextFloat() for float values is unknown
- JsonReader#nextInteger() for int32 values is unknown, it should be: JsonReader#nextInt()

**Comment by wing328** [comment]
If you've time, you may want to add the test to the Java (jersey 2) Petstore sample as well to see if you can uncover similar issues found with GSON.

**Comment by sdoeringNew** [comment]
Ok.

No new problems on the Jersey 2 sample. 😉
